### PR TITLE
[SPARK-12643][Build] Set lib directory for antlr

### PR DESCRIPTION
**Pull request by viirya**
JIRA: https://issues.apache.org/jira/browse/SPARK-12643

Without setting lib directory for antlr, the updates of imported grammar files can not be detected. So SparkSqlParser.g will not be rebuilt automatically.

Since it is a minor update, no JIRA ticket is opened. Let me know if it is needed. Thanks.

### CODE REVIEW

**Comment by hvanhovell** [comment]
@viirya isn't the lib directory the location where ANTLR writes ```.tokens``` files?

We should look into this though.

**Comment by viirya** [comment]
lib directory is "." by default, if you are asking for that?

ANTLR looks for imported grammar files (not generated .java files which are located with .tokens in same path now) in lib directory. Even if we set lib directory to where ANTLR writes .tokens files, it doesn't work because it can't find imported grammar files.

So ANTLR will not know imported grammar files are updated and will not rebuild SparkSqlParser.g.

**Comment by rxin** [comment]
We should definitely have a JIRA ticket for this under the parser umbrella ticket. Even though the change is small, if indeed a problem the impact is large.

@hvanhovell  I will let you review / decide if this is good to merge.

**Comment by hvanhovell** [comment]
@viirya I have been making a few changes to the grammar files, ANTLR is picking those changes up fine. I am using SBT.

How should I reproduce your problem?

**Comment by viirya** [comment]
@hvanhovell When you modify the imported grammar file such as IdentifiersParser.g, ANTLR will not go to rebuild the java files. I am using SBT to do test and build too.

**Comment by hvanhovell** [comment]
@viirya I did some digging.

The ```SparkSqlParser.g``` depends on 3 other grammars. When you touch the ```SparkSqlParser.g```or the ```SparkSqlLexer.g``` the ANTLR tool will detect the change and re-generate the parser/lexer. However, if you touch the dependent grammers, the ANTLR tool will not pick those if it can't find them. ANTLR can't find those files we haven't configurated the library path.

Your fix is the correct one.

**Comment by viirya** [comment]
@hvanhovell thanks for review.

File: project/SparkBuild.scala [hunk 1]
```diff
@@ -443,6 +443,7 @@ object Hive {
         val relGFilePath = (gFilePath relativeTo sourceDir).get.getPath
         log.info("ANTLR: Grammar file '%s' detected.".format(relGFilePath))
         antlr.addGrammarFile(relGFilePath)
+        antlr.setLibDirectory(gFilePath.getParent)
```
Comment (hvanhovell) [inline project/SparkBuild.scala]: Tiny thing: By doing this we set the library path multiple times to the same directory. There is unfortunately not an easy way out of this. The solution could be to fix the grammar dir, but that would make the build brittle.

Could you add a comment stating that we assume there is only one library directory?

Comment (hvanhovell) [inline project/SparkBuild.scala]: Brittle solution would be:

    val grammarDir = sourceDir / "org" / "apache" / "spark" / "sql" / "catalyst" / "parser"
    antlr.setLibDirectory(grammarDir.getPath)
    grammarFileNames.foreach { gFileName =>
        // We construct the entire path in order to make sure the file exists.
        val relGFilePath = ((grammarDir / gFileName) relativeTo sourceDir).get.getPath
        log.info("ANTLR: Grammar file '%s' detected.".format(relGFilePath))
        antlr.addGrammarFile(relGFilePath)
      }

Comment (viirya) [inline project/SparkBuild.scala]: Yes, I noticed that. You are correct.

I would prefer the current one even it sets the library path multiple times and assumes there is only one library directory.

If you and @rxin both think it is better, I can change it.

**Comment by hvanhovell** [comment]
LGTM

**Comment by rxin** [comment]
I've merged this. Thank.

**Pull request closed without merging**